### ISSUE TITLE
Unused variable optimization leads to `SyntaxError`s for IIFE-like function statements

### ISSUE THREAD

**Issue by sodatea** [rollup/rollup#4092]
- Rollup Version: 2.50.0
- Operating System (or Browser): macOS 11.3.1
- Node Version (if applicable): 14.17.0
- Link to reproduction (IMPORTANT, read below):

https://rollupjs.org/repl/?version=2.50.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMGElMjAlM0QlMjBmdW5jdGlvbiUyMCgpJTIwJTdCJTdEKCklMjAlM0YlMjBjb25zb2xlLmxvZygxKSUyMCUzQSUyMGNvbnNvbGUubG9nKDIpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

### Expected Behavior

Even though the variable name `a` is not referenced again in the scope, it is required to keep the module syntactically correct, so it should not be pruned.

### Actual Behavior

The variable declaration is pruned. The output code now throws a `SyntaxError` when executed.

---

This bug leads to some especially hard-to-locate issues when combined with other optimization techniques.

For example https://github.com/vitejs/vite/issues/3449
1. The `react-data-table-component` ES entry was first bundled and minimized without replacing `process.env.NODE_ENV`.
2. As `process.env.NODE_ENV` is unknown, the assigned value of the problematic function expression is not marked as unused, so it remains.
3. But as the code is also minimized, the function expression is transformed into a function statement (the parenthesis are removed) to save space. (But unlike IIFEs, a function statement MUST have a name)
4. When the package is depended on by the application, it is bundled again with Rollup (via Vite). This time, `process.env.NODE_ENV` is replaced by `"production"`. Therefore Rollup applies more optimization to the code, during which the assigned value becomes unused.
5. Then we encountered this bug.

The issue might occur sporadically. But when it happens, it is extremely hard to locate, because such patterns are most likely to be found in minimized codebases.

**Comment by kzc** [comment]
> This bug leads to some especially hard-to-locate issues when combined with other optimization techniques.

To isolate code generation errors in rollup, use the [`validate`](https://rollupjs.org/guide/en/#outputvalidate) CLI or config output option:

```
$ echo 'var a = function(){}() ? console.log(1) : console.log(2)' | rollup --validate --silent
(!) Chunk "-.js" is not valid JavaScript: Unexpected token (1:8).
-.js (1:8)
1: function(){}() ? console.log(1) : console.log(2);
           ^
function(){}() ? console.log(1) : console.log(2);
```

As `vite` appears to be built on top of rollup, it might be possible to add the output option `validate: true` to a `vite` config file. If this does not work, create an issue with the `vite` project to optionally support the rollup `validate` option.

**Comment by lukastaegert** [comment]
Fix at #4099
